### PR TITLE
tikv: fix infinite panic when GRPC transportReader is broken

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: refer https://github.com/pingcap/tidb/issues/18339

Problem Summary:

when panic like https://github.com/pingcap/tidb/issues/18339 happened, batchClient will recovery from panic without reCreate stream and meet panic again and again...

### What is changed and how it works?

What's Changed, How it Works:

although we didn't know how to broke transportReader, we can break infinite panic by covert panic to an error in `recv` method, then the error will trigger reCreate stream 

image

make it recovery from broken transportReader

### Related changes

- Need to cherry-pick to the release branch 4.0

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

```

```

Side effects

- n/a

### Release note 

- Fix infinite panic when GRPC transportReader is broken

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/18437)

### CODE REVIEW

File: store/tikv/client_batch.go [hunk 1]
```diff
@@ -538,6 +549,7 @@ func (c *batchCommandsClient) initBatchClient() error {
 		return errors.Trace(err)
 	}
 	c.client = streamClient
+	c.client.Context().Done()
```
Comment (tiancaiamao) [inline store/tikv/client_batch.go:552]: What's the purpose of this change?

https://pkg.go.dev/google.golang.org/grpc?tab=doc#ClientStream

```
	// It should not be called until after Header or RecvMsg has returned. Once
	// called, subsequent client-side retries are disabled.
	Context() context.Context
```

Comment (lysu) [reply store/tikv/client_batch.go]: orz， my fault

**Comment by tiancaiamao** [comment]
LGTM

**Comment by lysu** [comment]
@jackysp @coocood  PTAL thx

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
```

[2020-07-14T07:42:06.374Z] ==================

[2020-07-14T07:42:06.374Z] WARNING: DATA RACE

[2020-07-14T07:42:06.374Z] Write at 0x00c00066fd88 by goroutine 56:

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/executor.ResetContextOfStmt()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:1667 +0xe1d

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/session.(*session).ExecuteStmt()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1140 +0x1f1

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/session.(*session).Execute()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1081 +0x3ee

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/session.execRestrictedSQL()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:835 +0x15c

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/session.(*session).ExecRestrictedSQLWithContext()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:774 +0x25b

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:744 +0x92

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/statistics/handle.(*Handle).Update()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:162 +0x1eb

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:1099 +0x571

[2020-07-14T07:42:06.374Z] 

[2020-07-14T07:42:06.374Z] Previous read at 0x00c00066fd88 by goroutine 114:

[2020-07-14T07:42:06.374Z]   github.com/pingcap/tidb/executor.(*IndexLookUpExecutor).startIndexWorker.func1()

[2020-07-14T07:42:06.374Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/distsql.go:514 +0x477

[2020-07-14T07:42:06.374Z] 
```
